README.md
JasperEmbedSample
=================

Embedded Demo using visualize.js

Code: https://github.com/marianol/JasperEmbedSample

Author(s): [See list of Contributors](https://github.com/marianol/JasperEmbedSample/graphs/contributors)

Version: 1.2

##Requirements
- JasperReports Server 5.6
- A web server

##Installation

###Install the Sample
The sample contains a collection of HTML pages that needs to be tun form a WebServer. Each sample shows different aspects of the Visualize.js funtionality.
It is designed to be used with the samples that come with JasperServer 5.6.
The dashboard sample uses some extra reports which you can find in /JasperServerResources/dahsboard_reports.zip

- Unzip the distribution into your web server's web root
- Import /JapserServerResources/dahsboard_reports.zip to your JasperServer instance. [Check this link if you do not know how.](http://community.jaspersoft.com/documentation/jasperreports-server-administration-guide-beta/import-and-export-through-web-ui#import-export_2353750880_1044705)
- Go to http://<your-server>/JasperEmbedSample/ us the "Extra Samples" menu at the top to navigate trough the samples
- Enjoy!!!

###Installing the Dashboard Saver Servlet (Optional)
In the sample there is a dashboard builder that has the functionality to save a dashboard layout. To do this uses a java servlet running in Tomcat.
You can use the sample with out this servlet but you will not be able to save the dashboard layout.
- Configure your database connection in JasperServerResources/dashboardsaver/WEB-INF/web.xml
- Copy the folder of JasperServerResources/dashboardsaver into your Tomcat /webapps folder

##More Information and Resources
- http://community.jaspersoft.com/project/visualizejs
- http://community.jaspersoft.com/documentation/v56/jasperreports-server-programming-guide
- http://community.jaspersoft.com/documentation


LICENSE AND COPYRIGHT NOTIFICATION
==================================

 Copyright (C) 2005 - 2012 Jaspersoft Corporation - All rights reserved.

 Unless you have purchased a commercial license agreement from Jaspersoft,
 the following license terms apply:

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU Affero General Public License as
 published by the Free Software Foundation, either version 3 of the
 License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 GNU Affero  General Public License for more details.

 You should have received a copy of the GNU Affero General Public  License
 along with this program. If not, see <http://www.gnu.org/licenses/>.


changelog.md
#Change Log for Version 1.2

commit ce6e82f246030e847180fa481b117695cd20020c
Author: Mariano Luna 
Date:   Thu Jul 24 14:21:07 2014 -0300

    Moving the dahboardsaver to jasperServer Resources

commit 5d34568dcf118e28a60a21bba624d819f3f36c81
Author: Mariano Luna 
Date:   Thu Jul 24 14:17:13 2014 -0300

    New Read me file with the new additions.

commit 4492997f2db8c8cadbd5715e90f4fb988f680c50
Merge: 024cd36 feb16cd
Author: Mariano Luna 
Date:   Thu Jul 24 12:20:07 2014 -0300

    Merge branch 'refactoring'

commit 024cd3695fcb7120c201feb18bd92f626e0d940b
Author: Mariano Luna
Date:   Thu Jul 24 12:19:02 2014 -0300

    Cleanup.. simple.html was a leftover from the BIF implementation

commit feb16cdee2d95022520da1ecc9d80dd29ce2b1de
Author: Mariano Luna
Date:   Thu Jul 24 12:12:38 2014 -0300

    Fixed the Filter Stores function call from ReportPopUp

commit c4bf9af5f16c4c22bc3d02b9ce34f09cfe877ff5
Author: Gianluca Natali <gnatali@tibco.com>
Date:   Thu Jul 24 13:27:33 2014 +0200

    With this commit the following enanchements

     * Supporting variable configuration of jrs url (in the file config.js)
     * centralization of javascript in external files
     * reportPopUp sample now allow multiple pop up to be open and run indipendently
     * the pageSpecific.js contain all the specific javascript needed in various page. ther
     * new sample DashboardDesigner from Christoph Wurm (check it out!).

    This committ
    closes #6
    closes #5
    closes #4
    closes #1

commit ad054889adcca7bc8ea3c8b092cf25202621d636
Merge: 07007b4 4b28666
Author: Mariano Luna
Date:   Wed Jul 9 08:30:20 2014 -0500

    Merge branch 'Adding-Dashboard-Example'

commit 07007b4e305a355594067ff8fcc045c1b3e10d00
Author: Mariano Luna
Date:   Wed Jul 9 08:22:19 2014 -0500

    Clean up

commit 4b2866677fb9e29ce94dcd5de1d1ae43c0953b69
Author: Gene Arnold <HomeBase@Genes-MacBook-Pro-3.local>
Date:   Wed Jul 9 09:09:50 2014 -0400

    test

commit cc922671ec8a63d5b15fe935719862019c37bba8
Author: Gene Arnold <HomeBase@Genes-MacBook-Pro-3.local>
Date:   Wed Jul 9 09:06:01 2014 -0400

    Added Dashboard Sample


js/config.js
var config = {

	// URL of the JRS
	jrsUrl: "http://localhost:8080/jasperserver-pro",

	// JRS login username
	jrsUsername: "jasperadmin",

	// JRS login password
	jrsPassword: "jasperadmin",

	// JRS login organization
	jrsOrganization: "organization_1",
	
	dashBoardDesigner : {

		// The folder on JRS that will be searched for reports to add.
		folderUri: "/public/Dashboard_Reports", 

		// Indicates whether JRS should be searched for reports only
		// in the given folder (see option folderUri above) or also
		// in all sub-folders.
		recursive: false,

		// URL of the Dashboardsaver servlet
		dashboardSaver: "dashboard.do",

		// Initial height of the application, in px
		defaultApplicationHeight: 800,

		// Initial width of reports when added to a dashboard, in px
		defaultReportWidth: 400,

		// Initial height of reports when added to a dashboard, in px
		defaultReportHeight: 300
	}
}



js/visualizeHelper.js
/*
 * ========================================================================
 * visualizeHelper.js : v0.1.0
 * 
 * ========================================================================
 * Copyright 2014
 * Author: Mariano Luna (https://github.com/marianol)
 *
 * Unless you have purchased a commercial license agreement from Jaspersoft Inc., the following license terms apply:
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the 
 * GNU Affero General Public License as published by the Free Software Foundation, either version 3 
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public 
 * License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License 
 * along with this program. If not, see http://www.gnu.org/licenses/.
 * ======================================================================== 
 */
 
 var JRSClient; // Will be storing my JRS Client object here :)

 /*Loading Visualize javascript - with this we avoid to import the js file statically in page*/
 var myJRServer = {
	visualizeJS_libraryUrl : function() { 
		libraryUrl = config.jrsUrl + "/client/visualize.js"
		if( this.targetJrsUrl !== undefined){
			libraryUrl = libraryUrl + "?baseUrl=" + this.targetJrsUrl;
		}
		return libraryUrl;	
	}
 };

 //As soon as the script is loaded in the DOM we initialize the visualizeJS instance
$.getScript( myJRServer.visualizeJS_libraryUrl(), function() { 	
	launchVisualizeJs();
});

function launchVisualizeJs() {
	
	/*
	 Initialize Visualize.js config and JRSClient object placeholder

	Full Example:
	 myConfig = {
		 auth: {
			 name: "jasperadmin",
			 password: "jasperadmin",
			 organization:"organization_1",
			 locale: "en",
			 timezone: "Europe/Helsinki"
		 }
	 }
	 */
	jrsConfig =  {
		auth: {
			name: config.jrsUsername,
			password: config.jrsPassword,
			organization: config.jrsOrganization
		}
	};

	// Pass config to Visualize
	visualize.config(jrsConfig);
}


/**
 * Create and run report component with provided properties
 * @param {string} uri   - report properties
 * @param {string} container - div name to render the report
 * @param {Object} v - a visualize.js Jasper Client object (JrsClient)
 * @returns {Report} report - instance of Report generated
 */
function renderReport(uri, container, v) {
    return v.report({
        resource: uri,
        container: container,
        /*
        // I'll be setting this event handlers function as needed on the caller, see context-ic.html as an example
        events: {
            changeTotalPages: function(totalPages) {
                if ( typeof reportChangeTotalPages == 'function' ) {
                    reportChangeTotalPages(totalPages);
                }
            }
            reportCompleted: function(status, error) {
                if (status === "ready") {
                    // do something
                } else if (status === "failed") {
                    error && alert(error);
                }
            }
        },

        linkOptions: {
             events: {
                 "click"  : function(evt, link){

                     console.log(evt.target); // prints dom elem which represents link in report
                     console.log(link); // print link's data look in report data schema
                     }
             }
         },
        */
        error: function(err) {
                alert(err.message);
            }
        });

}

/**
 * Render standard Jasper Input Controls IC given the IC object
 * @todo extend this to handle all available IC Types
 *
 * @param {Options} inputParameters   -  inputControls instance
 *
 */
function renderStandardIC(inputParameters) {
    // Only working with single and multi selects now.
    $.each( inputParameters, function( id, inputControl ) {
        // console.log("IC # " + id + " - Label" + inputControl.label + " - Type: " + inputControl.type);

        var element;
        var elementProperties;

        $('#inputName').html(inputControl.label);
        /* Possible JRS IC Types:
         •	bool
         •	singleSelect
         •	singleSelectRadio
         •	multiSelectCheckbox
         •	multiSelect
         •	singleValue
         •	singleValueText
         •	singleValueNumber
         •	singleValueDate
         •	singleValueDatetime
         •	singleValueTime
         */
        switch(inputControl.type)
        {
            case 'singleSelect':
            case 'singleSelectRadio':
                element = '<select/>';
                elementProperties = {
                    'id': inputControl.id,
                    'name': inputControl.id,
                    'type': 'select-one'
                }
                break;
            case 'multiSelectCheckbox':
            case 'multiSelect':
                element = '<select/>'
                elementProperties = {
                    'id': inputControl.id,
                    'name': inputControl.id,
                    'type': 'select-multiple',
                    'multiple' : 'multiple'
                }
                break;
            default:
                element = '<input/>'
                elementProperties = {
                    'id': inputControl.id,
                    'type': inputControl.type
                }
        }
        var sel = $(element , elementProperties ).appendTo( '#inputOptions' );

        // @todo This should change to handle all types of Input Controls
        $.each(inputControl.state.options, function() {
            sel.append(
                        $( '<option />', {
                            'value': this.value,
                            'text': this.label
                        } )
                    );
        });
    });
}


function makeFrame() {
    ifrm = document.createElement("IFRAME");
    ifrm.setAttribute("src", "http://developerfusion.com/");
    ifrm.style.width = 640+"px";
    ifrm.style.height = 480+"px";
    document.body.appendChild(ifrm);

    /*
    // You can also use jQuery
         $('<iframe />', {
             name: 'myFrame',
             id:   'myFrame',
             ...
         }).appendTo('body');
     */


}



WEB-INF/classes/com/jaspersoft/vjsdashboard/DashboardSaver.java
package com.jaspersoft.vjsdashboard;

import java.io.IOException;
import java.io.PrintWriter;
import java.sql.Connection;
import java.sql.DatabaseMetaData;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;

import javax.servlet.ServletConfig;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;                                                                                   
import javax.servlet.http.HttpServletResponse;

public class DashboardSaver extends HttpServlet{

	private static final long serialVersionUID = -1952835428283006514L;
	
	private Connection connection = null;
	
	private String dbTableName = null;
	
	@Override
	public void init(ServletConfig servletConfig) throws ServletException {	
		String jdbcDriver = servletConfig.getInitParameter("jdbcDriver");
		String jdbcUrl = servletConfig.getInitParameter("jdbcUrl");
		String dbUsername = servletConfig.getInitParameter("dbUsername");
		String dbPassword = servletConfig.getInitParameter("dbPassword");
		dbTableName = servletConfig.getInitParameter("dbTableName");
		
		try {
			System.out.println(jdbcDriver);
            Class.forName(jdbcDriver);
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }

        try {
			connection = DriverManager.getConnection(jdbcUrl, dbUsername, dbPassword);
		} catch (SQLException e) {
			e.printStackTrace();
		}
	}
	
	@Override
	protected void doPost(HttpServletRequest request, HttpServletResponse response)
			throws ServletException, IOException {
		// To allow access from localhost
		response.addHeader("Access-Control-Allow-Origin", "*");
		
		String username = request.getParameter("username");
		String json = request.getParameter("json");
		
		try {
			createTableIfNotExists();
			
			PreparedStatement deleteSt = connection.prepareStatement("DELETE FROM " + dbTableName + " WHERE username = ?");
			deleteSt.setString(1, username);
			deleteSt.executeUpdate();
		
			PreparedStatement insertSt = connection.prepareStatement("INSERT INTO " + dbTableName + " (username, json) VALUES (?, ?)");
			insertSt.setString(1, username);
			insertSt.setString(2, json);
			insertSt.executeUpdate();
		} catch (SQLException e) {
			e.printStackTrace();
		}
	}

	private void createTableIfNotExists() throws SQLException {
		DatabaseMetaData metaData = connection.getMetaData();
		ResultSet rs = metaData.getTables(null, null, dbTableName, null);
		
		boolean noRows = !rs.first();
		if (noRows) {
			String createTableSql = "CREATE TABLE " + dbTableName + " (username varchar(20), json text)";
			Statement statement = connection.createStatement();
			statement.execute(createTableSql);
		}
	}

	@Override
	public void doGet(HttpServletRequest request, HttpServletResponse response)
    throws IOException{
		// To allow access from localhost
		response.addHeader("Access-Control-Allow-Origin", "*");
		
        PrintWriter out = response.getWriter();

        try {
	        PreparedStatement selectSt = connection.prepareStatement("SELECT json FROM " + dbTableName + " WHERE username = ?");
			selectSt.setString(1, request.getParameter("username"));
			ResultSet rs = selectSt.executeQuery();
			
			rs.next();
			out.write(rs.getString(1));
        } catch (SQLException e) {
			e.printStackTrace();
		}
    }
}

js/dashboardDesigner/vjsDashboard.js
// Code based on Gene Arnold's Dashboard Example
// https://github.com/GeneArnold/Dashboard-Example-using-Jaspersoft-Visualize.js-API



// Each tab has a unique number, running sequentially from 0 onward
var tabNum = 0;

function addTab() {
	$("#dashboard-container").append("<div id='tabs-" + tabNum + "' class='dashboard'></div>");
	$("#addTab").parent().before("<li class=\"tab\"><a href=\"#tabs-" + tabNum + "\">New Tab</a></li>");
	
	++tabNum;
				   
	$("#tabs").tabs("refresh");
	$("#tabs").tabs({ active: -2 });
          
	$newDashboard = $("#tabs-" + (tabNum - 1));
    $newDashboard.droppable();

	return $newDashboard.get();
}

function calcDashboardContainerHeight() {
	$("#dashboard-container")
		.height(
			$("#tabs").outerHeight()
			- $("#gallery").outerHeight()
			- $("#tabs-header").outerHeight());
}


// As Visualize.js's render method needs an id as a parameter we're creating here
// a variable that will be used and then incremented whenever a report is added.
var reportNum = 0;

// Using this function we are adding buttons to the screen to add and remove reports to the dashboard.
// Custom attributes are used to hold information about the status of the report's display and URI
// inside of JRS.
function listRepository(results) {
    $.each(results, function(index) {
        $("#gallery").append("<button class='addReportBtn' report-uri='"
								+ this.uri + "' report-title='" + this.label
								+ "' id='btn_" + index + "'>" + reportNameTrim(this.label)
								+ "</button>");

		calcDashboardContainerHeight();
    });
}

// Report name trim helper
function reportNameTrim(name) {
    if (name.length > 20) {
        var x = name.substring(0, 17) + '...';
        return x;
    } else {
        return name;
    }
}

function getCurrentDashboard() {
	return $(".dashboard[aria-hidden='false']");
}

function setReportHeight($reportContainer) {
	var $reportHeader = $reportContainer.children(".report-header");
	var $reportInputs = $reportContainer.children(".report-inputs");
	var $report = $reportContainer.children(".report");
	$report.height($reportContainer.outerHeight() - $reportHeader.outerHeight()
					- $reportInputs.outerHeight());

	// Fire window resize event to cause reports to be re-rendered without
	// calling JRS
	var evt = document.createEvent('UIEvents');
	evt.initUIEvent('resize', true, false, window, 0);
	window.dispatchEvent(evt);
}

function buildControl(control, $container) {
	if (typeof control.state.options != 'undefined'){
		function buildOptions(options) {
			var template = "<option>{value}</option>";
			return options.reduce(function (memo, option) {
				return memo + template.replace("{value}", option.value);
			}, "")
		}

		$container.append("<tr><td><label>" + control.label + "</td>"
		+ "<td><select param-id=" + control.id + " multiple='true'><option></option>"
		+ buildOptions(control.state.options) + "</select></td></tr>");
	}else{
		$container.append("<tr><td><label>" + control.label + "</td>"
		+ "<td><input type='text' param-id=" + control.id + "></td></tr>");
	}	
}

function createReportToolbar() {
	function toggleReportInputs($reportContainer) {
		$reportContainer.css("height", "auto");
		var $reportInputs = $reportContainer.find(".report-inputs");
		if ($reportInputs.css("height").length > 0) {
			$reportInputs.css("height", $reportInputs.height());
		}
		$reportInputs.slideToggle();
	}

    var $inputControls =$("<div class='inputControls'>Y</div>")
		.click(function() {
			var $reportContainer = $(this).closest(".report-container");
			var $reportHeader = $reportContainer.children(".report-header");
			var $report = $reportContainer.children(".report");

			if ($reportContainer.find(".report-inputs").length > 0) {
				toggleReportInputs($reportContainer);
			} else {
				var $setButton = $("<button>Set</button>")
					.click(function() {
						var params = new Object();

						$("[param-id]").each(function(index, select) {
							var $paramValue = $(select).val();
							if ($paramValue !== null & select.localName == 'select') {
								params[$(select).attr("param-id")] = $(select).val();
							}else if($paramValue !== null & select.localName == 'input'){
								var v = [];
								v.push($(select).val());
								params[$(select).attr("param-id")] = v;
							}
						});

						$report.data("jsreport").params(params).run();
					});
				
				$reportContainer.css("height", "auto");
				var $loading = $("<div class='report-inputs'></div>")
					.append("<span class='loading'>Loading...</span>")
					.insertAfter($reportHeader)
					.hide()
					.slideDown();

				var $reportInputs = $("<div class='report-inputs'></div>")
					.append("<table></table>")
					.append($setButton);

				JRSClient.inputControls({
					resource: $report.attr("report-uri"),
					success: function (controls) {
						$loading
							.slideUp(function() {
								$loading.remove();
								$reportInputs
									.insertAfter($reportHeader)
									.hide();

								toggleReportInputs($reportContainer);
							});

						$.each(controls, function(index, control) {
							buildControl(control, $reportInputs.find("table"));
						});

						$reportInputs.find("select")
							.chosen({
								width: ($reportHeader.width() * 0.8) + "px"
							})
							.on("change", function() {
								$reportContainer.css("height", "auto");
								$reportInputs.css("height", "auto");

								/* Because of the Set button that is floated. */
								$reportInputs.css("padding-bottom", "5ex");
							});

						toggleReportInputs($reportContainer);
					},
					error: function (err) {
						console.log(err);
					}
				});
			}
		});


	var $minimize = $("<div class='minimize'>_</div>")
		.click(function() {
			var $reportContainer = $(this).closest(".report-container");
			var $reportHeader = $reportContainer.children(".report-header");
			var $report = $reportContainer.children(".report");

			if ($reportContainer.height() > $reportHeader.outerHeight() + 50) {
				$reportContainer
					.animate({
						height: "-=" + $report.outerHeight()
					});
			} else {
				$reportContainer
					.animate({
						height: "+=" + $report.outerHeight()
					});
			}
		});

	var $maximize = $("<div class='maximize'>&#9633;</div>")
		.click(function() {
			var $reportContainer = $(this).closest(".report-container");

			if (!$reportContainer.hasClass("maximized")) {
				$reportContainer.addClass("maximized");
			
				$reportContainer.data({
					origLeft: $reportContainer.css("left"),
					origTop: $reportContainer.css("top"),
					origWidth: $reportContainer.css("width"),
					origHeight: $reportContainer.css("height")
				});

				$reportContainer.css({
					position: "relative",
					left: 0,
					top: 0,
					width: "100%",
					height: "100%"
				});
			} else {
				$reportContainer.removeClass("maximized");

				$reportContainer.css({
					position: "absolute",
					left: $reportContainer.data("origLeft"),
					top: $reportContainer.data("origTop"),
					width: $reportContainer.data("origWidth"),
					height: $reportContainer.data("origHeight")
				});
			}
			
			setReportHeight($reportContainer);
		});

	var $close = $("<div class='close'>X</div>")
		.click(function() {
			$(this).closest(".report-container").remove();
		});

	return $("<div class='toolbar'></div>")
		.append($inputControls, $minimize, $maximize, $close);
}

function addReport(reportURI, reportTitle, start_x, start_y, size_x, size_y, reportParams) {
	var $reportContainer = $("<div></div>")
		.css({
			position:"absolute",
			left: start_x,
			top: start_y,
			width: size_x,
			height: size_y
		})
		.addClass("report-container")
		.draggable({
			containment: "parent"
		})
		.resizable({
			stop: function() {
                setReportHeight($(this));
			}
		})
		.appendTo(getCurrentDashboard());

	var $reportHeader = $("<div></div>")
		.addClass("report-header")
		.append(createReportToolbar())
		.append("<span class='report-title'>" + reportTitle + "</span>")
		.appendTo($reportContainer);

	var $report = $("<div></div>")
		.addClass("report")
		.attr({
			id: "report-" + reportNum++,
			"report-uri": reportURI
		})
		.appendTo($reportContainer);

	setReportHeight($reportContainer);
    
	if (reportParams === undefined) {
		reportParams = new Object();
	}

    var report = JRSClient.report(
        {
            resource: reportURI,
			params: reportParams,
            container: "#" + $report.attr("id"),
            events:
                {
                    reportCompleted: function(status) {}
                },
            error:
                function(err) {
                     console.log(err.message);
                }
        }
    );
    
    $report.data("jsreport", report);
}

// Create a JSON structure that holds all the tabs, reports and their locations and sizes
// and sends it to be saved
function saveArrangement() {
	var json = new Object();
    json.tabs = new Array();

	// Build JSON
	$(".dashboard").each(function() {
        var tab = new Object();
		tab.reports = new Array();

		$.each($(this).children(".report-container"), function(index, reportContainer) {
            var report = new Object();
			
			report.uri = $(reportContainer).children(".report").data("jsreport").resource();
			report.params = $(reportContainer).children(".report").data("jsreport").params();
			report.title = $(reportContainer).children(".report-header").children(".report-title").text();
			report.start_x = $(reportContainer).css("left");
			report.start_y = $(reportContainer).css("top");
			report.size_x = $(reportContainer).css("width");
			report.size_y = $(reportContainer).css("height");

			tab.reports.push(report);
        });

		json.tabs.push(tab);
	});

	// Send to server
	$.post(config.dashBoardDesigner.dashboardSaver, 
			{ username: $("#username input").val(), json: JSON.stringify(json) });
}

function restoreArrangement() {
	$.getJSON(config.dashBoardDesigner.dashboardSaver,
				{ username: $("#username input").val() },
				function(data) {
					$(".tab, .dashboard, .ui-dialog").remove();

					$.each(data.tabs, function(index, tab) {
						var uiTab = addTab();

						$.each(tab.reports, function(index, report) {
							addReport(report.uri, report.title, report.start_x, report.start_y,
									  report.size_x, report.size_y, report.params);
						});
					});
				}
			);
}

function setupTabsAndButtons() {
	var tabs = $("#tabs").tabs({
		beforeActivate: function(event, ui) {
			if (ui.newTab.context.id === "addTab") {
				event.preventDefault();
				addTab();
			}
		}
	});
  
	// Default first tab
	addTab();

	$("#tabs").height(config.dashBoardDesigner.defaultApplicationHeight);
	calcDashboardContainerHeight();

	$("#tabs").resizable({
		resize: function() {
			calcDashboardContainerHeight();
		}
	});

	$("#save").click(saveArrangement);
	$("#restore").click(restoreArrangement);
}

jQuery.fn.extend({
	isInsideOf: function(elem) {
		if (this.length > 0) {
			if (this.offset().left > elem.offset().left
				&& this.offset().left + this.outerWidth() < elem.offset().left + elem.outerWidth()
				&& this.offset().top > elem.offset().top
				&& this.offset().top + this.outerHeight() < elem.offset().top + elem.outerHeight()
			) {
				return true;
			} else {
				return false;
			}
		}
	}
});

function setupDragAndDrop() {
	// Drag and drop according to http://stackoverflow.com/a/14445678/235203
	var click = false;
	var top = null;
	var left = null;

	$(document).bind('mousemove', function (e) {
		if (click == true) {
			$('#drag').css({
				left: e.pageX,
				top: e.pageY
			});
		}
	});
	    
	$('html').click(function() {
		if ($("#drag").isInsideOf(getCurrentDashboard())) {
			click = false;

			getCurrentDashboard().removeClass("ui-state-highlight");

			addReport($("#drag").attr("report-uri"), $("#drag").attr("report-title"),
						$("#drag").offset().left - $("#tabs").offset().left,
						$("#drag").offset().top - $("#tabs").offset().top,
						config.dashBoardDesigner.defaultReportWidth, config.dashBoardDesigner.defaultReportHeight);

			$("#drag").remove();
		}
	});

	$("#gallery").on("mousedown", ".addReportBtn", function(e) {
		var $newReportContainer = $("<div id='drag'></div>")
			.css({
				width: config.dashBoardDesigner.defaultReportWidth,
				height: config.dashBoardDesigner.defaultReportHeight, 
				position: "fixed",
				left: e.pageX,
				top: e.pageY
			})
			.addClass("report-container")
			.draggable({
				revert: "invalid"
			})
			.attr({
				"report-uri": $(e.target).attr("report-uri"),
				"report-title": $(e.target).attr("report-title")
			})
			.appendTo($("html"));

		left: e.pageX;
		top = e.pageY;

		click = true;

		getCurrentDashboard().addClass("ui-state-highlight");
	});
}

function DashboardDesignerPageJS(){

	this.init = function(windowObj){	
		initializePage(windowObj, initialize); 	
	}
	
	// Get a list of all reports the logged in user can access and display them
	function initialize() {
		JRSClient.resourcesSearch(
			{
				server: config.jrsUrl,
				folderUri: config.dashBoardDesigner.folderUri,
				recursive: config.dashBoardDesigner.recursive,
				success: listRepository,
				types: ["reportUnit"],
				error:
					function (err) {
						alert(err);
					}
			}
		);
	}

}


$(document).ready(function() {
	setupTabsAndButtons();
	setupDragAndDrop();
});
